perception/autoware_tensorrt_bevdet/src/nvjpegdecoder.cpp

```diff
@@ -0,0 +1,299 @@
+    CHECK_NVJPEG(nvjpegDecoderCreate(
+      share_param.nvjpeg_handle, NVJPEG_BACKEND_DEFAULT, &params[i].nvjpeg_decoder));
+    CHECK_NVJPEG(nvjpegDecoderStateCreate(
+      share_param.nvjpeg_handle, params[i].nvjpeg_decoder, &params[i].nvjpeg_decoupled_state));
+
+    CHECK_NVJPEG(
+      nvjpegBufferPinnedCreate(share_param.nvjpeg_handle, NULL, &params[i].pinned_buffers[0]));
+    CHECK_NVJPEG(
+      nvjpegBufferPinnedCreate(share_param.nvjpeg_handle, NULL, &params[i].pinned_buffers[1]));
+    CHECK_NVJPEG(
+      nvjpegBufferDeviceCreate(share_param.nvjpeg_handle, NULL, &params[i].device_buffer));
+
+    CHECK_NVJPEG(nvjpegJpegStreamCreate(share_param.nvjpeg_handle, &params[i].jpeg_streams[0]));
+    CHECK_NVJPEG(nvjpegJpegStreamCreate(share_param.nvjpeg_handle, &params[i].jpeg_streams[1]));
+
+    CHECK_NVJPEG(
+      nvjpegDecodeParamsCreate(share_param.nvjpeg_handle, &params[i].nvjpeg_decode_params));
+  }
+}
+
+void destroy_decoupled_api_handles(
+  std::vector<decode_params_t> & params, share_params & share_param)
+{
+  for (size_t i = 0; i < params.size(); i++) {
+    CHECK_NVJPEG(nvjpegDecodeParamsDestroy(params[i].nvjpeg_decode_params));
+
+    CHECK_NVJPEG(nvjpegJpegStreamDestroy(params[i].jpeg_streams[0]));
+    CHECK_NVJPEG(nvjpegJpegStreamDestroy(params[i].jpeg_streams[1]));
+    CHECK_NVJPEG(nvjpegBufferPinnedDestroy(params[i].pinned_buffers[0]));
+    CHECK_NVJPEG(nvjpegBufferPinnedDestroy(params[i].pinned_buffers[1]));
+    CHECK_NVJPEG(nvjpegBufferDeviceDestroy(params[i].device_buffer));
+
+    CHECK_NVJPEG(nvjpegJpegStateDestroy(params[i].nvjpeg_decoupled_state));
+    CHECK_NVJPEG(nvjpegDecoderDestroy(params[i].nvjpeg_decoder));
+  }
+}
+
+void release_buffers(std::vector<nvjpegImage_t> & ibuf)
+{
+  for (size_t i = 0; i < ibuf.size(); i++) {
+    for (size_t c = 0; c < NVJPEG_MAX_COMPONENT; c++)
+      if (ibuf[i].channel[c]) CHECK_CUDA(cudaFree(ibuf[i].channel[c]));
+  }
+}
+
+int get_img(
+  const uchar * d_chanR, int pitchR, const uchar * d_chanG, int pitchG, const uchar * d_chanB,
+  int pitchB, size_t width, size_t height, uchar * img)
+{
+  size_t step = width * height;
+
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 0 * step, static_cast<size_t>(width), d_chanR, static_cast<size_t>(pitchR), width, height,
+    cudaMemcpyDeviceToDevice));
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 1 * step, static_cast<size_t>(width), d_chanG, static_cast<size_t>(pitchG), width, height,
+    cudaMemcpyDeviceToDevice));
+  CHECK_CUDA(cudaMemcpy2D(
+    img + 2 * step, static_cast<size_t>(width), d_chanB, static_cast<size_t>(pitchB), width, height,
+    cudaMemcpyDeviceToDevice));
+
+  return EXIT_SUCCESS;
+}
+
+void decode_single_image(
+  const std::vector<char> & img_data, nvjpegImage_t & out, decode_params_t & params,
+  share_params & share_param, double & time)
+{
+  CHECK_CUDA(cudaStreamSynchronize(params.stream));
+  cudaEvent_t startEvent = NULL, stopEvent = NULL;
+  float loopTime = 0;
+  // cudaEventBlockingSync
```

Comment:
The use of `cudaEventCreate`, `cudaEventRecord`, `cudaEventSynchronize`, and `cudaEventElapsedTime` for timing is good. However, the comments suggest potential inconsistencies (`// cudaEventBlockingSync`, `//TODO new add`, `// time = 1.0;`). Ensure that the timing mechanism is consistent and accurate across all decoding paths.   Also consider using a more sophisticated timing mechanism that avoids the overhead of CUDA events if precise timing is not essential.